test/kubernetes_unittest.cc

```diff
@@ -1951,4 +1966,52 @@ TEST_F(KubernetesTestFakeServerThreeWatchRetriesNodeLevelMetadata,
   EXPECT_TRUE(server->TerminateAllStreams(time::seconds(3)));
 }
 
+namespace {
+class MockSleeper : public Sleeper {
+ public:
+  MOCK_METHOD1(SleepFor, void(double seconds));
+};
+
+// Polls health_checker until it has at least one unhealthy component.
+// Returns false if newer timestamp not found after 3 seconds (polling
+// every 100 millis).
+bool WaitForUnhealthyComponents(const HealthChecker& health_checker) {
+  for (int i = 0; i < 30; i++){
+    if (health_checker.UnhealthyComponents().size() > 0) {
+      return true;
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  return false;
+}
+}
+
+TEST_F(KubernetesTestFakeServerMaxThreeFailuresNodeLevelMetadata,
+       KubernetesUpdaterBackoffOnFailures) {
+  // Create a fake server representing the Kubernetes master, but
+  // don't register streams for the nodes or pods watch paths, so they
+  // will return 404s.
+  //
+  // This will cause the updater to backoff 3 times and then give up.
+  server->SetResponse("/api/v1/nodes?limit=1", "{}");
+  server->SetResponse("/api/v1/pods?limit=1", "{}");
+
+  MetadataStore store(*config);
+  HealthChecker health_checker(*config, store);
+  // Note: These expectations don't check the order in which the
+  // SleepFor() calls are invoked.  Each is called twice, once each
+  // for the nodes & pods watchers.
+  auto sleeper = std::unique_ptr<MockSleeper>(new MockSleeper());
+  EXPECT_CALL(*sleeper, SleepFor(1.5)).Times(2);
```

Comment:
Is there a reason why checking that `Sleep` was called with appropriate arguments via a mock is better than faking the passage of time and verifying that certain events happened at certain times?

Reply:
I couldn't find anyway of faking out std::this_thread::sleep_for(). Got any suggestions?